internal/provider/resource_network.go

```diff
@@ -790,3 +880,163 @@ func resourceNetworkDelete(ctx context.Context, d *schema.ResourceData, meta int
 	}
 	return diags
 }
+
+// CreateAccessNodeCapacity is used when a capacity contract is needed for wg access.
+//
+// an already used farm is preferred to create the capacity contract on. if no farm is eligible, a random farm is selected.
+func (k *NetworkDeployer) CreateAccessNodeCapacity() (contractID uint64, err error) {
+	var minFreeIPs *uint64
+	var dedicated *bool
+	*minFreeIPs = 1
+	*dedicated = false
+
+	// preferebly choose a farm used with the capacity contracts, if not, choose a random one
+	farmID, err := k.getPreferrableFarm()
+	if err != nil {
+		return 0, fmt.Errorf("couldn't get a farm to create a capacity contract on for wg access. %w", err)
+	}
+	resources := substrate.Resources{
+		HRU: 0,
+		SRU: 0,
+		CRU: 0,
+		MRU: 0,
+	}
+	// log chosen farm
+	log.Printf("farm %d was chosen to create capacity contract on for wg access. %d", farmID)
+
+	policy := substrate.WithCapacityPolicy(resources, substrate.NodeFeatures{IsPublicNode: true})
+	contractID, err = k.APIClient.substrateConn.CreateCapacityReservationContract(k.APIClient.identity, uint32(farmID), policy, nil)
+	if err != nil {
+		return 0, fmt.Errorf("couldn't create capacity contract for wg access. %w", err)
+	}
+	log.Printf("capacity contract id %d was created for wg access", contractID)
+	return contractID, nil
+}
+
+// getPreferrableFarm gets a farm to create a capacity contract on.
+// this farm is preferrably one of the farms already used in provided capacity contracts.
+// if non is eligible, a random eligible farm is chosen
+func (k *NetworkDeployer) getPreferrableFarm() (uint64, error) {
+	var freeIP *uint64
+	var dedicated *bool
+	*freeIP = uint64(1)
+	*dedicated = false
+	for _, id := range k.CapacityIDs {
+		contract, err := k.APIClient.substrateConn.GetContract(id)
+		if err != nil {
+			return 0, fmt.Errorf("couldn't get contract with id %d", id)
+		}
+		nodeID := uint32(contract.ContractType.CapacityReservationContract.NodeID)
+		node, err := k.APIClient.substrateConn.GetNode(nodeID)
+		if err != nil {
+			return 0, fmt.Errorf("couldn't get node with id %d. %w", id, err)
+		}
+		var farmID *uint64
```

Comment:
please review the usage of pointer semantics here, you can pass the address of the value farmID instead of creating a pointer to it, and for filtering if you want to represent absence of that filter pass nil instead.